src/libs/ajax/User.ts

```diff
@@ -0,0 +1,398 @@
+import _ from 'lodash/fp';
+import * as qs from 'qs';
+import {
+  authOpts,
+  fetchBond,
+  fetchEcm,
+  fetchOk,
+  fetchOrchestration,
+  fetchRex,
+  fetchSam,
+  jsonBody,
+} from 'src/libs/ajax/ajax-common';
+import { getConfig } from 'src/libs/config';
+import { getTerraUser } from 'src/libs/state';
+import * as Utils from 'src/libs/utils';
+
+export interface SamUserRegistrationStatusResponse {
+  userSubjectId: string;
+  userEmail: string;
+  enabled: boolean;
+}
+
+export interface SamUserTosStatusResponse {
+  userInfo: {
+    userSubjectId: string;
+    userEmail: string;
+  };
+  enabled: {
+    ldap: boolean;
+    allUsersGroup: boolean;
+    google: boolean;
+  };
+}
+
+export interface SamUserTosComplianceStatusResponse {
+  userId: string;
+  userHasAcceptedLatestTos: boolean;
+  permitsSystemUsage: boolean;
+}
+
+export interface OrchestrationUserProfileResponse {
+  userId: string;
+  keyValuePairs: { key: any; value: any }[];
+}
+
+interface OrchestrationUserPreferLegacyFireCloudResponse {
+  preferTerra: boolean;
+  preferTerraLastUpdated: number;
+}
+
+interface NihDatasetPermission {
+  name: string;
+  authorized: boolean;
+}
+interface OrchestrationNihStatusResponse {
+  linkedNihUsername: string;
+  datasetPermissions: NihDatasetPermission[];
+  linkExpireTime: number;
+}
+
+interface BondFenceUrlResponse {
+  url: string;
+}
+
+interface BondFenceStatusResponse {
+  issued_at: Date;
+  username: string;
+}
+
+interface EcmLinkAccountResponse {
+  externalUserId: string;
+  expirationTimestamp: Date;
+  authenticated: boolean;
+}
+
+export interface SamInviteUserResponse {
+  userSubjectId: string;
+  userEmail: string;
+}
+interface RexFirstTimestampResponse {
+  timestamp: Date;
+}
+
+// TODO: Remove this as a part of https://broadworkbench.atlassian.net/browse/ID-460
+const getFirstTimeStamp = Utils.memoizeAsync(
+  async (token): Promise<RexFirstTimestampResponse> => {
+    const res = await fetchRex('firstTimestamps/record', _.mergeAll([authOpts(token), { method: 'POST' }]));
+    return res.json();
+  },
+  { keyFn: (...args) => JSON.stringify(args) }
+) as (...args: any[]) => RexFirstTimestampResponse;
```

Comment:
This doesn't really accept any arguments, it accepts a token. Also, it returns a Promise for the response.

Does this work?
```suggestion
) as (token: string) => Promise<RexFirstTimestampResponse>;
```

I know `memoizeAsync` doesn't have types, so it might be difficult to get it to cooperate. If nothing else, 
```suggestion
) as unknown as (token: string) => Promise<RexFirstTimestampResponse>;
```